examples/scripts/spin.py

```diff
@@ -0,0 +1,221 @@
+    for split in ["train", "test"]:
+        processed_datasets[split] = processed_datasets[split].rename_columns(
+            {"text_prompt": "prompt", "text_real": "real", "text_generated": "generated"}
+        )
+
+    torch_dtype = (
+        model_args.torch_dtype if model_args.torch_dtype in ["auto", None] else getattr(torch, model_args.torch_dtype)
+    )
+    quantization_config = get_quantization_config(model_args)
+
+    model_kwargs = dict(
+        revision=model_args.model_revision,
+        trust_remote_code=model_args.trust_remote_code,
+        attn_implementation=model_args.attn_implementation,
+        torch_dtype=torch_dtype,
+        use_cache=False if training_args.gradient_checkpointing else True,
+        device_map=get_kbit_device_map() if quantization_config is not None else None,
+        quantization_config=quantization_config,
+    )
+
+    model = model_args.model_name_or_path
+
+    ref_model = model
+    ref_model_kwargs = model_kwargs
+
+    if model_args.use_peft is True:
+        ref_model = None
+        ref_model_kwargs = None
+
+    #########################
+    # Instantiate spin trainer
+    #########################
+    spin_trainer = SPINTrainer(
+        model,
+        ref_model,
+        model_init_kwargs=model_kwargs,
+        ref_model_init_kwargs=ref_model_kwargs,
+        args=training_args,
+        beta=training_args.beta,
+        train_dataset=processed_datasets["train"],
+        eval_dataset=processed_datasets["test"],
+        tokenizer=tokenizer,
+        max_length=training_args.max_length,
+        max_prompt_length=training_args.max_prompt_length,
+        peft_config=get_peft_config(model_args),
+    )
+
+    if script_args.do_generate:
+        prompt_train_ds = raw_datasets["train"]
+        prompt_test_ds = raw_datasets["test"]
+        generation_config = GenerationConfig(
+            do_sample=True,
+            temperature=script_args.temperature,
+            max_new_tokens=script_args.max_new_tokens,
+            pad_token_id=tokenizer.eos_token_id,
+        )
+
+    ###############
+    # Training loop
+    ###############
+    if training_args.do_train:
+        train_result = spin_trainer.train()
+        metrics = train_result.metrics
+        spin_trainer.log_metrics("train", metrics)
+        spin_trainer.save_metrics("train", metrics)
+        spin_trainer.save_state()
+        spin_trainer.save_model(training_args.output_dir)
+        spin_trainer.push_to_hub()
+
+        print(f"Training completed! Model pushed to {spin_trainer.hub_model_id}")
+
+    if script_args.do_generate:
+        print(f"Generating completions for {len(prompt_train_ds)} training examples")
+        train_completions = spin_trainer.generate(prompt_train_ds, "prompt", generation_config, batch_size=16)
+        test_completions = spin_trainer.generate(prompt_test_ds, "prompt", generation_config, batch_size=16)
```

Comment:
If the generation only occurs in the end, wdyt about doing a subprocess call on a vllm/tgi script to generate instead?